### ISSUE TITLE
Could not write to xterm

### ISSUE THREAD

**Issue by edwin-egalite** [git-for-windows/git#124]
Using Git for Windows 2.3.7.1x64 on Windows 8.1 x64, when doing Git pull, I get the following error:

```
$ git pull
warning: Could not write to xterm
fatal: could not read Password for 'http://edwin@xxx.xxx.xxx.xxx:port': No error
```

Add below line to global config fixed this:

```
askpass = C:/Program Files/Git/mingw64/libexec/git-core/git-gui--askpass
```

**Comment by dscho** [comment]
Git Bash or Git CMD? What is the value of the `TERM` environment variable (both inside and outside of *Git Bash*)?

**Comment by edwin-egalite** [comment]
Git Bash, never use Git in CMD. No value in environment variable was set during installation.

**Comment by dscho** [comment]
> > What is the value of the TERM environment variable (both inside and outside of Git Bash)?
> No value in environment variable was set during installation.

I was asking for the actual values in your setup. I was asking because I cannot reproduce, and I was prying for information that would allow me to reproduce the problem here. I still do not have enough information to be able to do that, and hence I have no chance at fixing your problem for you. Yet.

**Comment by nalla** [comment]
Since [terminal.c#L109](https://github.com/git-for-windows/git/blob/master/compat/terminal.c#L109) was not thrown, I imply that `MSYS_TTY_HANDLES` is set.

Maybe the handle in `MSYS_TTY_HANDLES` is not valid anymore in your setup. Is that error *always* reproduceable?

**Comment by dscho** [comment]
@edwin-egalite please make it easier to help you by providing the information @nalla and I need to know to help.

**Comment by edwin-egalite** [comment]
@nalla Yes, this error is always reproduceable. I do a clean re-installation, and clone a repo from Bitbucket. When doing Git pull, it will throw ```Could not write to xterm``` warning and the bash become not responding.

I do not sure how to check whether or not the handle in ```MSYS_TTY_HANDLES``` is functioning in my setup.

@dscho Would appreciate if you could guide me on providing the information you need for debugging.

**Comment by dscho** [comment]
@edwin-egalite please note that I would have really appreciated a [Minimal, Complete and Verifiable Example](http://stackoverflow.com/help/mcve). Such an example is essential to debug and fix the issue, and since debugging and fixing bugs is a pretty time consuming task, it would have seemed fair to me if you would have spent some time to make it easier on me.

This is the MCVE with which I came up:

```sh
mkdir 123
cd 123
git init
git pull https://edwin-egalite@github.com/invalid/repository
```

Please note that you can work around that in the meantime thusly:

```sh
git fetch
git merge @{u}
```

**Comment by nalla** [comment]
With the MVCE provided, it seems ´echo´ stays off after that too...

**Comment by dscho** [comment]
Yep, that is correct. The `echo` stays off because the `stty -echo </dev/tty` succeeds, but writing to the handle we got does not. The handle is most likely invalid due to `git pull` being a shell script that calls into the MinGW executable `git fetch` which then hands off to the MinGW executable `git remote-https` which in turn wants to read from the console, and somehow somewhere along the line the handle gets stale.

My idea to fix this which I want to try out today is to call out to `sh -c "cat >/dev/tty && stty -echo </dev/tty && read </dev/tty"` with `stdin` and `stdout` redirected to pipes, then write the prompt to, and read back the password from the process. After that, *even if the command failed*, I would call `sh -c "stty echo </dev/tty"` to make sure that the `echo` is on again.

This strategy *should* work because MSys2 processes access the same MSys2 heap, and they should be able to do so even if there is a non-MSys2 process in the call chain (as indicated by `stty -echo </dev/tty` switching off the `echo` even if it was called from the non-MSys2 executable `git-remote-https.exe`).

Let's see how that goes.

**Comment by dscho** [comment]
> Let's see how that goes.

*Almost* as I had written it. Solved as per d56d2f8.

**Comment by nalla** [comment]
For further quality assurance I tested `git push --set-upstream <github-https-remote> <branch>` . Worked fine.
